discovery/mdns.go

```diff
@@ -215,21 +222,42 @@ func (r *MDNS) read(ctx context.Context, conn *net.UDPConn) {
 		for addr, name := range entries {
 			if isValidName(name) {
 				if r.addrs == nil {
-					r.addrs = map[string][]string{}
-					r.names = map[string][]string{}
+					r.addrs = map[string]mdnsEntry{}
+					r.names = map[string]mdnsEntry{}
 				}
 				name := absDomainName([]byte(name))
 				h := []byte(name)
 				lowerASCIIBytes(h)
 				key := absDomainName(h)
-				r.addrs[addr] = appendUniq(r.addrs[addr], name)
-				r.names[key] = appendUniq(r.names[key], addr)
+				addEntry(r.addrs, addr, name)
+				addEntry(r.names, key, addr)
+				for len(r.addrs) > mdnsMaxEntries {
+					r.removeOldestEntry()
+				}
```

Comment:
@rs The issue still seems to exist when testing using the build you gave me in this comment: https://github.com/nextdns/nextdns/issues/886#issuecomment-1890922498

At present, nextdns is reporting over 5,000 discovered mdns hosts:

```
# nextdns discovered|jq '.mdns | length'
5081
```

The name of my HA install is `hass`,  and here are just a sample of the mdns entries:

```
  "hass9956.local.": [
    "fd23:62ab:6fcf:b69b:fe98:6ac0:12cd:bde0",
    "fe80::fefa:6a68:d08f:4e20"
  ],
  "hass9965.local.": [
    "fe80::1c9a:80a1:12dc:e39f",
    "10.200.0.3"
  ],
  "hass9967.local.": [
    "fe80::1c9a:80a1:12dc:e39f",
    "10.200.0.3"
  ],
  "hass9971.local.": [
    "fe80::fefa:6a68:d08f:4e20",
    "fd23:62ab:6fcf:b69b:fe98:6ac0:12cd:bde0",
    "192.168.55.3"
  ],
```


Looking at the code I'm commenting on, I believe that you should be checking the length of `r.names`. In my environment, the number of addresses seems to be stable at 66 entries:

```
# nextdns discovered|jq '.mdns | .[] | flatten[]'|sort | uniq -c|wc -l
66
```

However, I do see that the HA addresses are associated with a large number of mdns hostnames:
```
# nextdns discovered|jq '.mdns | .[] | flatten[]'|sort | uniq -c|sort -n|tail -n 6
      3 "10.200.197.10"
   2379 "10.200.0.3"
   2419 "192.168.55.3"
   3678 "fe80::1c9a:80a1:12dc:e39f"
   3859 "fd23:62ab:6fcf:b69b:fe98:6ac0:12cd:bde0"
   3860 "fe80::fefa:6a68:d08f:4e20"
```